### PR TITLE
fix(headless): allowed the constant query to be unset

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/KIT-450

The initial value of the constant query was an empty string. Unfortunately, while an empty string is often interpreted as "not set" for other values, for the constant query it meant "no query", or "don't filter", which meant a user explicitly setting the constant query to an empty string was interpreted as being equivalent to the user having never set the constant query at all. For most types of queries, this is fine, since nobody would want a non-empty default, but for the constant query this caused issues with the Tab controller, which would always overwrite an explicit "no query" with the constant query of any Tab controller whose "isActive" initial state was true.

In a nutshell, the following behaviour was possible even in a page with `SearchParameterManager` that properly reads and writes the URL:
1. Create a page with three tabs. "All" (cq=""), "Messages" (cq="...", initialState.isActive=true)
2. Load the page, the "Messages" tab is active
3. Select the "All" tab
4. Refresh the page, the "Messages" tab is active again

### CODE REVIEW

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts [hunk 1]
```diff
@@ -119,8 +119,9 @@ function getCq(state: Partial<SearchParametersState>) {
   }
 
   const cq = state.advancedSearchQueries.cq;
-  const shouldInclude = cq !== getAdvancedSearchQueriesInitialState().cq;
-  return shouldInclude ? {cq} : {};
+  const shouldInclude =
+    cq !== null && cq !== getAdvancedSearchQueriesInitialState().cq;
```
Comment (btaillon) [inline packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:123]: I'm uncertain whether I should or shouldn't include this null check since `getAdvancedSearchQueriesInitialState().cq` is null too.

The purpose of this condition is essentially just not to include `cq` if nothing ever set it.

File: packages/headless/src/controllers/tab/headless-tab.ts [hunk 2]
```diff
@@ -80,7 +80,7 @@ export function buildTab(
     'buildTab'
   );
 
-  if (initialState.isActive) {
+  if (initialState.isActive && engine.state.advancedSearchQueries.cq === null) {
```
Comment (btaillon) [inline packages/headless/src/controllers/tab/headless-tab.ts:83]: This is the most important part of this PR, but it has a flaw: if somebody intends to change which tab is currently active by creating a new controller whose `isActive` is true, it won't work anymore. Personally, I don't think that's much of a problem since in order for that to happen, a developer would need to be attempting to manage tabs themselves instead of reading the intended `state.isActive` and calling the intended `select`.

File: packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts [hunk 3]
```diff
@@ -11,6 +11,8 @@ export interface AdvancedSearchQueriesState {
 }
 
 export const getAdvancedSearchQueriesInitialState: () => AdvancedSearchQueriesState = () => ({
-  cq: '',
+  cq: null,
   aq: '',
 });
+
+export const getConstantQueryDefaultValue = () => '';
```
Comment (btaillon) [inline packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts:18]: I don't really like this, but my imagination kinda fails me on that one... Should this be a constant or a function? What should it be named? Does it belong somewhere else? Alternatively, it could be removed and replaced with `?? ""` whenever anything checks for null.

File: packages/headless/src/controllers/tab/headless-tab.ts (hunk 2 above)
Comment (samisayegh) [reply packages/headless/src/controllers/tab/headless-tab.ts:83]: Agreed, there should only ever be one Tab controller with `isActive`. If a dev wants to make another tab active, they would have to remove the `isActive` on the other instance.

As for the condition, I think we should have this inside the reducer itself, and have the controller dispatch a special `registerAdvancedSearchQueries` or `registerConstantQuery` action.

**Comment by ThibodeauJF** [comment]
There is no difference between the cq and aq or q really, it is how we use it with the tabs which is different. We shouldn't make that difference explicit in the state by setting cq's type to `string | null`.

It also looks a bit weird to have `#cq=` in the url... I'm not so sure there is a "clean" solution here...

I like Sami's proposition for a possible fix of shouldInclude instead of the current `null` design for now. For it to be "clean", it comes back to have cq sets instead of a single string in the state perhaps... which is breaking, we were thinking of revisiting tabs anyways in the future.

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts (hunk 1 above)
Comment (btaillon) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:123]: I prefer not comparing the `cq` to the initial request since the `initialState` feels pretty representative of what should be the initial constant query, and one or more requests could theoretically be sent before this happens (whether because of an implementation or our own changes).

I like the idea of a `initialConstantQuery` reducer though, that seems like it would work well in any situation and be easier to fix if something comes up. I'll add that.

File: packages/headless/src/app/search-app-reducers.ts [hunk 4]
```diff
@@ -42,6 +43,7 @@ export const searchAppReducers: ReducersMapObject<SearchAppState> = {
   categoryFacetSearchSet: categoryFacetSearchSetReducer,
   query: queryReducer,
   advancedSearchQueries: advancedSearchQueriesReducer,
+  defaultConstantQueryFilter: defaultConstantQueryFilterReducer,
```
Comment (samisayegh) [inline packages/headless/src/app/search-app-reducers.ts:46]: I believe this should be part of the `advancedSearchQueries` reducer. It does not make sense to split the logic imo. A register action would set the initial cq value and the `cq` provided no initial cq has been set.

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts [hunk 5]
```diff
@@ -43,7 +46,7 @@ export type SearchParameterManager = ReturnType<
 export type SearchParameterManagerState = SearchParameterManager['state'];
 
 export function buildSearchParameterManager(
-  engine: Engine<Partial<SearchParametersState>>,
+  engine: Engine<Partial<SearchAppState>>,
```
Comment (samisayegh) [inline packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:49]: This interface change should not be necessary if we place the new property inside of the `advancedSearchQueries` slice.

File: packages/headless/src/controllers/tab/headless-tab.test.ts [hunk 6]
```diff
@@ -38,10 +39,42 @@ describe('Tab', () => {
...
+        },
+      };
+      engine.state.defaultConstantQueryFilter.actualValueWasChanged = false;
+      initTab();
+
       const action = updateAdvancedSearchQueries({cq: expression});
       expect(engine.actions).toContainEqual(action);
     });
 
+    it("doesn't call #registerConstantQuery if isActive is true and actaulValueWasChanged is true", () => {
```
Comment (samisayegh) [inline packages/headless/src/controllers/tab/headless-tab.test.ts:62]: There should be no `actaulValueWasChanged` state in the controller. This should all be handled by the reducer.

Controllers are just meant to simplify interacting with the engine. One should be able to get the same behaviour by dispatching actions.

When we starting adding stateful logic to controllers, it now means you either must use the controller, or you have to reimplement the same checks when using the actions. When the logic is in the reducer, you can do either and get the same experience.

File: packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts [hunk 7]
```diff
@@ -1,16 +1,53 @@
+export interface AdvancedSearchQueriesDefaultFiltersState {
+  /**
+   * The initial cq filter (e.g., `((q AND aq) OR dq) AND cq).
+   */
+  cq: string;
+
+  /**
+   * The initial aq filter (e.g., `((q AND aq) OR dq) AND cq).
+   */
+  aq: string;
```
Comment (btaillon) [inline packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts:10]: I added support to `aq` as well since it makes the behaviour of `AdvancedSearchQueries` more consistent across both, therefor less confusing. If you prefer, I can remove support for the default value for `aq`.

File: packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts [hunk 8]
```diff
@@ -1,16 +1,53 @@
...
+   * Whether the aq filter was manually set.
+   *
+   * If the aq was manually set and the aq is registered, the aq will not be overriden by the default filter.
+   */
+  aqWasSet: boolean;
+
+  /**
+   * The initial filters meant to be used as default values for the cq filter and aq filter.
+   */
+  defaultFilters: AdvancedSearchQueriesDefaultFiltersState;
```
Comment (btaillon) [inline packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts:41]: I'm still uncertain about using the term `default`...

File: packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts [hunk 9]
```diff
@@ -1,16 +1,53 @@
...
    * The cq filter (e.g., `((q AND aq) OR dq) AND cq).
    */
   cq: string;
 
+  /**
+   * Whether the cq filter was manually set.
+   *
+   * If the cq was manually set and the cq is registered, the cq will not be overriden by the default filter.
+   */
+  cqWasSet: boolean;
```
Comment (btaillon) [inline packages/headless/src/features/advanced-search-queries/advanced-search-queries-state.ts:24]: `cqWasSet` is important, since it's possible that `#restoreSearchParameters` may be called before `#registerAdvancedSearchQueries`, in which case we don't want to override it with the default.

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts [hunk 10]
```diff
@@ -108,19 +107,17 @@ function getAq(state: Partial<SearchParametersState>) {
...
 function getCq(state: Partial<SearchParametersState>) {
   if (state.advancedSearchQueries === undefined) {
     return {};
   }
 
-  const cq = state.advancedSearchQueries.cq;
-  const shouldInclude = cq !== getAdvancedSearchQueriesInitialState().cq;
-  return shouldInclude ? {cq} : {};
+  const {cq, cqWasSet} = state.advancedSearchQueries;
+  return cqWasSet ? {cq} : {};
```
Comment (btaillon) [inline packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: The replacement of `shouldInclude` with `cqWasSet` means that manually setting the `cq`, even manually setting it to the default value, will include it in the search parameters.

I have a slight preference for having this behaviour, but I don't mind keeping the old behaviour if you prefer.

Comment (samisayegh) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: I just got an idea looking at the new condition. Since it always adds the `cq` as long as it was set, it is no longer hiding the `cq` parameter when the value is equal to the default.

Isn't this in the same as having `shouldInclude` always set to `true`? If so, then it seems it would be simpler to just always include the `cq` value, and remove the conditional all together. This is in effect what the new code is doing?

**Review by samisayegh — COMMENTED** [review]
Looking at the new code, I think it is the same as always including the `cq`. Could this simplification remove the need for all the additional properties?

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts (hunk 10 above)
Comment (ThibodeauJF) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: all urls using headless will contain `cq=` which can be pretty cryptic.

**Comment by ThibodeauJF** [comment]
I'm not sure we need both a `wasSet` and a `default` in the state, one of the do with good management of the reducer & `shouldInclude` should do it, no?

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts (hunk 10 above)
Comment (btaillon) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: Making the change only include `cq` at all times would cause the `initialState` property of the `Tab` controller to become useless. Thus, loading something like `https://somewebsite/search` would always select the "all" tab.

Comment (samisayegh) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: Ideally we would only include the `cq` in the url if it's value is different from the initial cq value.

With the current changes is that the case?
Won't `cqWasSet` be `true` the moment a tab is [updated](https://github.com/coveo/ui-kit/pull/522/files#diff-afab97fef7d9ccf209a66b2972b76192ef6a00c27118f06f34710f42b6e36dccR23)?

If the initial tab is `All`, I click on `youtube`, then go back to `All`. I believe the cq will be in the url whereas before it would not have been?

**Comment by btaillon** [comment]
> I'm not sure we need both a `wasSet` and a `default` in the state, one of the two with good management of the reducer & `shouldInclude` should do it, no?

Including only the `default` means that the tab **must** be registered before the search parameters manager calls `restoreSearchParameters`, which even in the React sample turned out not to be the case. Otherwise, whichever `tab` is being set in the URL would always be overwritten by the default.

Including only `wasSet` however looks like it may work, so long as we're ok with the behaviour of this PR (never removing the parameter if `cq` was explicitly set).

File: packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts (hunk 10 above)
Comment (btaillon) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: Yes, the last scenario you described is the case with this PR, this is what I tried to describe in my [first comment](https://github.com/coveo/ui-kit/pull/522#discussion_r582230466).

It may be possible to remove the `default` state value and keep this behaviour. However, if we want the `cq` to be able to disappear from the URL when it's equal to the default value, we need to keep it.

Comment (samisayegh) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: If we want to stay consistent with the other url parameters, we would only add the `cq` to the url if the current `cq` expression is different from the initial value.

Comment (btaillon) [reply packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts:120]: Then I'll add back the default and compare against it
